hbase-client/src/main/java/org/apache/hadoop/hbase/client/trace/IpcClientSpanBuilder.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.hadoop.hbase.client.AsyncConnectionImpl;
+import org.apache.hadoop.hbase.net.Address;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.RpcSystem;
+import org.apache.hadoop.hbase.trace.TraceUtil;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.hbase.thirdparty.com.google.protobuf.Descriptors;
+
+/**
+ * Construct {@link Span} instances originating from the client side of an IPC.
+ */
+@InterfaceAudience.Private
+public class IpcClientSpanBuilder implements Supplier<Span> {
+
+  private String name;
+  private final Map<AttributeKey<?>, Object> attributes = new HashMap<>();
+
+  public IpcClientSpanBuilder(
+    final Supplier<String> connectionStringSupplier,
+    final Supplier<User> userSupplier
+  ) {
+    // TODO: this constructor is a hack used by AbstractRpcClient because it does not have access
+    //  to the AsyncConnectionImpl within which it resides. Use this for now, until we come back
+    //  and plumb through the instance.
+    ConnectionSpanBuilder.populateConnectionAttributes(attributes, connectionStringSupplier,
+      userSupplier);
+  }
+
+  @Override
+  public Span get() {
+    return build();
+  }
+
+  public IpcClientSpanBuilder setMethodDescriptor(final Descriptors.MethodDescriptor md) {
+    // it happens that `getFullName` returns a string in the $package.$service format required by
+    // the otel RPC specification. Use it for now; might have to parse the value in the future.
+    final String packageAndService = md.getService().getFullName();
+    final String method = md.getName();
+    this.name = packageAndService + "/" + method;
+    populateMethodDescriptorAttributes(attributes, md);
+    return this;
+  }
+
+  public IpcClientSpanBuilder setRemoteAddress(final Address remoteAddress) {
+    attributes.put(NET_PEER_NAME, remoteAddress.getHostName());
+    attributes.put(NET_PEER_PORT, (long) remoteAddress.getPort());
+    return this;
+  }
+
+  @SuppressWarnings("unchecked")
+  public Span build() {
+    final SpanBuilder builder = TraceUtil.getGlobalTracer()
+      .spanBuilder(name)
+      // TODO: what about clients embedded in Master/RegionServer/Gateways/&c?
+      .setSpanKind(SpanKind.CLIENT);
+    attributes.forEach((k, v) -> builder.setAttribute((AttributeKey<? super Object>) k, v));
+    return builder.startSpan();
+  }
+
+  /**
+   * Static utility method that performs the primary logic of this builder. It is visible to other
+   * classes in this package so that other builders can use this functionality as a mix-in.
+   * @param attributes the attributes map to be populated.
+   * @param md the source of the RPC attribute values.
+   */
+  static void populateMethodDescriptorAttributes(
+    final Map<AttributeKey<?>, Object> attributes,
+    final Descriptors.MethodDescriptor md
+  ) {
+    // it happens that `getFullName` returns a string in the $package.$service format required by
```

Comment:
Redundant code?

Reply:
Not really. Reading fields out of the `Method` instance is redundant, yes, but it's just member variable access. The cautionary comment is repeated because it pertains to both code sections.